File: guests/ocaml/scrollto.ml
(* The scroll-to scene, OCaml port — guests/rust/scrollto.rs,
   tools/scenes/scrollto.steps: the app scrolls a list of messages to a
   row by key (docs/scroll-to-plan.md), opening at the newest one before
   the first layout, jumping to one on a click, staying put on a key no
   row holds, and following its own send. *)

open Kaya_app

type message = { text : string } [@@deriving kaya_gen]

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
      let messages = collection_of message_record in
      let count = signal Scalar.Str "60 messages" in
      (* The For's own container is what a scroll_to_row addresses
         (docs/scroll-to-plan.md S1): [each] returns it. *)
      let list = ref None in
      let sent = ref 60 in

      let send () =
        incr sent;
        let key = Printf.sprintf "m%d" !sent in
        insert_record messages (Key.str key)
          { text = Printf.sprintf "message %d" !sent };
        write count (Printf.sprintf "%d messages" !sent);
        scroll_to_row (Option.get !list) (Key.str key)
      in

      let root =
        column
          [
            label ~bind:count ~a11y_id:"count";
            row
              [
                button ~text:"jump" ~a11y_id:"jump" ~on_click:(fun () ->
                    scroll_to_row (Option.get !list) (Key.str "m10"));
                button ~text:"nowhere" ~a11y_id:"nowhere"
                  ~on_click:(fun () ->
                    scroll_to_row (Option.get !list) (Key.str "m999"));
                button ~text:"send" ~a11y_id:"send" ~on_click:send;
              ];
            scroll ~grow:1.0
              [
                (fun () ->
                  let w =
                    each (record_handle messages)
                      (fun () -> Tpl.(label ~bind_field:message_text ()))
                      ()
                  in
                  set_a11y_id w "messages";
                  list := Some w;
                  w);
              ];
          ]
          ()
      in
      mount root;

      for i = 1 to 60 do
        insert_record messages
          (Key.str (Printf.sprintf "m%d" i))
          { text = Printf.sprintf "message %d" i }
      done;
      scroll_to_row (Option.get !list) (Key.str "m60"));

  exit (run app)

File: guests/swift/scrollto.swift
// The scroll-to scene, Swift port — guests/rust/scrollto.rs,
// tools/scenes/scrollto.steps. The app scrolls a list of messages to a
// row by key (docs/scroll-to-plan.md), opening at the newest one before
// the first layout, jumping to one on a click, staying put on a key no
// row holds, and following its own send.

import Foundation
import Kaya

struct Message: KayaGen {
    var text: String
}

KayaApp.run { app in
    app.build { tx in
        let messages = messageCollection(tx)
        let count = tx.signal(.str("60 messages"))
        var sent = 60
        // The For's own container is what a scroll_to_row addresses
        // (docs/scroll-to-plan.md S1): the eliminator hands it back.
        var list: KayaWidget!

        let root = tx.column { root in
            tx.setA11yId(tx.label(bind: count), "count")
            tx.row { _ in
                tx.setA11yId(
                    tx.button("jump", onClick: { t in
                        t.scrollToRow(list, .str("m10"))
                    }), "jump")
                tx.setA11yId(
                    tx.button("nowhere", onClick: { t in
                        t.scrollToRow(list, .str("m999"))
                    }), "nowhere")
                tx.setA11yId(
                    tx.button("send", onClick: { t in
                        sent += 1
                        messages.insert(
                            t, .str("m\(sent)"), Message(text: "message \(sent)"))
                        t.write(count, .str("\(sent) messages"))
                        t.scrollToRow(list, .str("m\(sent)"))
                    }), "send")
            }
            tx.scroll(grow: 1) { _ in
                list = messageEach(tx, messages) { row in
                    row.label(row.text)
                }
                tx.setA11yId(list, "messages")
            }
            return root
        }
        tx.mount(root)

        for i in 1...60 {
            messages.insert(tx, .str("m\(i)"), Message(text: "message \(i)"))
        }
        tx.scrollToRow(list, .str("m60"))
    }
}
